libs/remote_storage/src/s3_bucket.rs

```diff
@@ -569,64 +624,142 @@ impl RemoteStorage for S3Bucket {
+                matches!(vd, VerOrDelete::DeleteMarker(_))
+            );
+
+            hm_vd
+                .entry(key)
+                .or_default()
+                .push((vd, last_modified, version_id));
+        }
+        for (key, versions) in hm_vd {
+            let (last_vd, last_last_modified, _version_id) = versions.last().unwrap();
+            if last_last_modified > &&done_if_after {
+                tracing::trace!("Key {key} has version later than done_if_after, skipping");
+                continue;
+            }
+            // the version we want to restore to.
+            let version_to_restore_to =
+                match versions.binary_search_by_key(&timestamp, |tpl| *tpl.1) {
+                    Ok(v) => v,
+                    Err(e) => e,
+                };
+            if version_to_restore_to == versions.len() {
+                tracing::trace!("Key {key} has no changes since timestamp, skipping");
+                continue;
+            }
+            let mut do_delete = false;
+            if version_to_restore_to == 0 {
+                // All versions more recent, so the key didn't exist at the specified time point.
+                tracing::trace!(
+                    "All {} versions more recent for {key}, deleting",
+                    versions.len()
+                );
+                do_delete = true;
+            } else {
+                match &versions[version_to_restore_to - 1] {
+                    (VerOrDelete::Version(_), _last_modified, version_id) => {
+                        tracing::trace!("Copying old version {version_id} for {key}...");
+                        // Restore the state to the last version by copying
+                        let source_id =
+                            format!("{}/{key}?versionId={version_id}", self.bucket_name);
+
+                        self.client
+                            .copy_object()
+                            .bucket(self.bucket_name.clone())
+                            .key(key)
+                            .copy_source(source_id)
+                            .send()
+                            .await?;
+                    }
+                    (VerOrDelete::DeleteMarker(_), _last_modified, _version_id) => {
+                        do_delete = true;
                     }
                 }
-                Err(e) => {
-                    return Err(e.into());
+            };
+            if do_delete {
+                if matches!(last_vd, VerOrDelete::DeleteMarker(_)) {
+                    // Key has since been deleted (but there was some history), no need to do anything
+                    tracing::trace!("Key {key} already deleted, skipping.");
+                } else {
+                    tracing::trace!("Deleting {key}...");
+
+                    let oid = ObjectIdentifier::builder().key(key.to_owned()).build()?;
+                    self.delete_oids(kind, &[oid]).await?;
```

Comment:
Should we batch these up?

Reply:
Thinking about this more, as the version based recovery is meant for large tenants with a lot of keys, batching and possibly even concurrent execution might be needed for that reason. (My earlier thinking was based on api availability :))

Reply:
We could do the copies first, while collecting all the deletions, and then do the deletions in a later step. copies cannot be batched up.

Reply:
Good for me as a follow-up.